go/vt/vtgate/vindexes/cfc.go

```diff
@@ -0,0 +1,326 @@
+		return v, nil
+	}
+	n := len(v)
+	m := len(vind.offsets)
+	// if we are not working on a prefix, the key has to have all the components,
+	// that is, it has to be longer than the largest offset.
+	if !prefix && n < vind.offsets[m-1] {
+		return nil, fmt.Errorf("insufficient size for cfc vindex %s. need %d, got %d", vind.name, vind.offsets[m-1], n)
+	}
+	prev := 0
+	offset := 0
+	buf := new(bytes.Buffer)
+	for _, offset = range vind.offsets {
+		if n < offset {
+			// the given prefix doesn't align with the component boundaries,
+			// return the hashed prefix that's currently available
+			return buf.Bytes(), nil
+		}
+
+		if _, err := buf.Write(vind.hash(v[prev:offset])); err != nil {
+			return nil, err
+		}
+		if n == offset {
+			break
+		}
+		prev = offset
+	}
+	if offset < n {
+		if _, err := buf.Write(vind.hash(v[offset:n])); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+// Verify returns true if ids maps to ksids.
+func (vind *CFC) Verify(_ VCursor, ids []sqltypes.Value, ksids [][]byte) ([]bool, error) {
+	out := make([]bool, len(ids))
+	for i := range ids {
+		v, err := vind.computeKsid(ids[i].ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = bytes.Equal(v, ksids[i])
+	}
+	return out, nil
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *CFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		v, err := vind.computeKsid(id.ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = key.DestinationKeyspaceID(v)
+	}
+	return out, nil
+}
+
+// PrefixVindex switches the vindex to prefix mode
+func (vind *CFC) PrefixVindex() SingleColumn {
+	return vind.prefixVindex
+}
+
+func addOne(value []byte) []byte {
+	n := len(value)
+	overflow := true
+	for i := n - 1; i >= 0; i-- {
+		if value[i] < 255 {
+			value[i]++
+			overflow = false
+			break
+		} else {
+			value[i] = 0
+		}
+	}
+	if overflow {
+		return nil
+	}
+	return value
+}
+
+type prefixCFC struct {
+	name string
+	cfc  *CFC
+}
+
+func (vind *prefixCFC) String() string {
+	return vind.name
+}
+
+func (vind *prefixCFC) Cost() int {
+	return 1
+}
+
+func (vind *prefixCFC) IsUnique() bool {
+	return false
+}
+
+func (vind *prefixCFC) NeedsVCursor() bool {
+	return false
+}
+
+// Verify returns true if ids maps to ksids.
+func (vind *prefixCFC) Verify(vc VCursor, ids []sqltypes.Value, ksids [][]byte) ([]bool, error) {
+	// prefixCFC is only active for 'LIKE' expr. Verify is used by 'INSERT' so we re-use CFC.Verify()
+	return vind.cfc.Verify(vc, ids, ksids)
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *prefixCFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		value := id.ToBytes()
+		// we only consider the string up to the first whild char as the prefix.
+		p := bytes.IndexByte(value, '%')
```

Comment:
In your implementation, you allow the input value to be arbitrary bytes. With that, you could run into trouble, because if the input is arbitrary, you could potentially have an escaped `%` in the input.

Reply:
This is a good point. I added some logic to deal with escaped `%`.